pkg/apis/stable/v1alpha1/fleetautoscaler.go

```diff
@@ -0,0 +1,112 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// +genclient
+// +genclient:noStatus
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// FleetAutoScaler is the data structure for a FleetAutoScaler resource
+type FleetAutoScaler struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   FleetAutoScalerSpec   `json:"spec"`
+	Status FleetAutoScalerStatus `json:"status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// FleetAutoScalerList is a list of Fleet Scaler resources
+type FleetAutoScalerList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	Items []FleetAutoScaler `json:"items"`
+}
+
+// FleetAutoScalerSpec is the spec for a Fleet Scaler
+type FleetAutoScalerSpec struct {
+	FleetName string `json:"fleetName"`
+	// MaxReplicas is the maximum amount of replicas that the fleet may have. 
+	// If zero, it is ignored. 
+	// If non zero, it must be bigger than both MinReplicas and BufferSize
+	MaxReplicas int32 `json:"maxReplicas"`
+	
+	// MinReplicas is the minimum amount of replicas that the fleet must have
+	// If zero, it is ignored. 
+	// If non zero, it must be smaller than MaxReplicas and bigger than BufferSize
+	MinReplicas int32 `json:"minReplicas"`
+	
+	// BufferSize defines how many replicas the autoscaler tries to have ready all the time
+	// Must be bigger than 0
+	// Note: by "ready" we understand in this case "non-allocated"; this is done to ensure robustness 
+	//       and computation stability in different edge case (fleet just created, not enough capacity in the cluster etc)
+	BufferSize int32 `json:"bufferSize"`
+}
+
+// FleetAutoScalerStatus needs doc
+type FleetAutoScalerStatus struct {
+	// Fleet is the copy of fleet we're attached to. It is updated each tick
+	Fleet *Fleet
```

Comment:
I'm not a fan of this - it's a lot of redundant information to be updating all the time.

It's also not necessary, since the Fleet is in local cache as well.

Looking at the HorizontalPodAutoscaler - the `desiredReplicas` and `currentReplicas` would likely be enough?
What extra information do we need here?

(As I write this, I realise I've got the full game server in the allocation - I'm now also questioning that decision!)

Reply:
Yes I agree we don't need the Fleet object, I just didn't know what to put in its stead.
How about a subresource compatible with 'kubectl scale'.... that could be used to scale bufferSize?

Reply:
I think `desiredReplicas` and `currentReplicas` would be good to put in the status, it gives enough information to those need to look into what is going on with the Autoscaler (and we have precedent with the HorizontalPodAutoscaler)

We should probably also have some events - but wanted to take this for a spin first before making suggestions.

Once we have #329 going, then there's an interesting question on how 'scale' should work wiith autoscalers -- let's definitely keep that in mind!

Reply:
done